src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+        // to historical storage as needed
+        if (curNotSynched || prevNotSynched)
+        {
+            if (prev)
+            {
+                // Move the formerly second latest shard to
+                // historical storage
+                if (keepShard(*prev) && separateHistoricalPath)
+                {
+                    moveShard(*prev);
+                }
+
+                prev = boost::none;
+            }
+
+            if (cur)
+            {
+                // The formerly latest shard is now
+                // the second latest
+                if (cur == latestShardIndex - 1)
+                {
+                    prev = cur;
+                }
+
+                // The formerly latest shard is no
+                // longer a 'recent' shard
+                else
+                {
+                    // Move the formerly latest shard to
+                    // historical storage
+                    if (keepShard(*cur) && separateHistoricalPath)
+                    {
+                        moveShard(*cur);
+                    }
+                }
+
+                cur = boost::none;
+            }
+        }
+    }
+}
+
+boost::optional<bool>
+DatabaseShardImp::expectNewShard(
+    std::uint32_t shardIndex,
+    std::uint32_t numHistoricalShards,
+    std::lock_guard<std::mutex> const& lock)
+{
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    auto const boundaryIndex = getShardBoundaryIndex(lock);
+    auto const isHistoricalShard = shardIndex < boundaryIndex;
+
+    // Check shard count and available storage space
+    if (isHistoricalShard && numHistoricalShards >= maxHistoricalShards_)
+    {
+        JLOG(j_.error()) << "maximum number of historical shards reached";
+        canAdd_ = false;
+        return boost::none;
+    }
+    if (!sufficientStorage(1, isHistoricalShard, lock))
+    {
+        JLOG(j_.error()) << "insufficient storage space available";
+        canAdd_ = false;
+        return boost::none;
+    }
+
+    return isHistoricalShard && !historicalPaths_.empty();
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath()
+{
+    std::lock_guard lock(mutex_);
+    return chooseHistoricalPath(lock);
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath(std::lock_guard<std::mutex> const&)
+{
+    // If not configured with separate historical paths,
+    // use the main path (dir_) by default.
+    if (historicalPaths_.empty())
+        return dir_;
+
+    std::array<boost::filesystem::path, 1> historicalShardPath;
+    std::vector<boost::filesystem::path> potentialPaths;
+
+    std::for_each(
+        historicalPaths_.begin(),
+        historicalPaths_.end(),
+        [this, &potentialPaths](boost::filesystem::path const& path) {
+            if (boost::filesystem::space(path).available >= avgShardFileSz_)
+                potentialPaths.push_back(path);
+        });
```

Comment:
Consider using a range based for loop instead. It's shorter, and more common than `for_each` in our code.

Reply:
Done